src/qibolab/platforms/multiqubit.py

```diff
@@ -45,14 +139,44 @@ def execute_pulse_sequence(self, sequence: PulseSequence, nshots=None):
             if "readout" in roles[name]:
                 if not instrument_pulses[name].is_empty:
                     if not instrument_pulses[name].ro_pulses.is_empty:
+                        if all([pulse.serial in changed for pulse in instrument_pulses[name].ro_pulses]):
+                            # FIXME: for precision sweep in resonator spectroscopy
+                            # change necessary to perform precision sweep
+                            # TODO: move this to instruments (ask Alvaro)
+                            # TODO: check if this will work with multiplex
+                            for sequencers in self.instruments[name]._sequencers.values():
+                                for sequencer in sequencers:
+                                    sequencer.pulses = instrument_pulses[name].ro_pulses
                         results = self.instruments[name].acquire()
                         existing_keys = set(acquisition_results.keys()) & set(results.keys())
                         for key, value in results.items():
                             if key in existing_keys:
                                 acquisition_results[key].update(value)
                             else:
                                 acquisition_results[key] = value
-        return acquisition_results
+
+        # change back the frequency of the pulses
+        for name in self.instruments:
+            roles[name] = self.settings["instruments"][name]["roles"]
+            if "readout" in roles[name]:
+                instrument_pulses[name] = sequence.get_channel_pulses(*self.instruments[name].channels)
+                for pulse in instrument_pulses[name]:
+                    if pulse.serial in changed:
+                        pippo = acquisition_results[pulse.serial]
+                        # if abs(pulse.frequency) > 300e6:
+                        pulse.frequency += self.get_lo_readout_frequency(pulse.qubit)
+                        acquisition_results[pulse.serial] = pippo
+            if "control" in roles[name]:
+                instrument_pulses[name] = sequence.get_channel_pulses(*self.instruments[name].channels)
+                for pulse in instrument_pulses[name]:
+                    if pulse.serial in changed:
+                        pulse.frequency += self.get_lo_drive_frequency(pulse.qubit)
+
+        for key in ro_pulses:
+            data[key] = ExecutionResults(
+                acquisition_results[key][0], acquisition_results[key][1], np.array(acquisition_results[key][2])
+            )
```

Comment:
@AleCandido if you choose to use a container how exactly should we modify this line?
Something like this?
```python
result = ExecRes(acquisition_results[key][0], acquisition_results[key][1], np.array(acquisition_results[key][2])
data[key] = ExecContainer(result)

```

Reply:
In this case, `ExecRes` is the type of the individual point, so we need to make an array first:
```py
ires = acquisition_results[key][0]
ar = np.empty(len(ires), dtype=ExecRes)
ar[`i`] = ires
ar[`q`] = acquisition_results[key][1]
ar[`sample`] = acquisition_results[key][2]
data[key] = ExecContainer(ar)
```
https://stackoverflow.com/a/17458107/8653979

However, you can also wrap the first lines in a class method, and just construct there:
```py
class ExecContainer:
    ...
    
    @classmethod
    def from_components(cls, is_, qs, samples):
        ar = np.empty(len(is_), dtype=ExecRes)
        ar[`i`] = is_
        ar[`q`] = qs
        ar[`sample`] = samples
        
        return cls(ar)
        
data[key] = ExecContainer.from_components(*acquisition_results[key])
```

Reply:
What is the advantage of this approach?

Reply:
That since your basic object is in practice a 3-tuple, you can keep working on it with all the NumPy functions and arrays, for every dimension you wish.

Otherwise, you have to decide how many array dimensions you put inside, and the outside ones will become lists operated with for-loops.
Let's say that you decide that you have a 1-d array structure typically, so you make a class with 3 arrays inside. Then at some point you have a collection of them (that was already happening), so you need a list of Python objects, or an NumPy array of Python objects containing NumPy arrays, and it is useless as well (because Python objects can't be summed, multiplied, ... by NumPy, unless they implement NumPy array interface, that is much more complicate, and definitely unneeded).
With structured arrays instead, you can simply decide to have a 2-dim structure array, simply as you had a 1-dim array with `dtype=ExecRes`, now you will have a 2-dim array with `dype=ExecRes`, and it scales.

If the object were more complex, I would not suggest to go this way. But the object is essentially an 1-dim array of length 3, but with inhomogeneous types. Structured arrays are perfect to represent them.